### PR TITLE
Check for MusicBrainz Release Id during Scan (#227)

### PR DESCRIPTION
**Pull request by majora2007**
#### Database Migration
NO

#### Description
Check if an album has a MusicBrainz Release Id during Import/Scan and check if a match in DB. If so, bypass metadata lookup against LidarrAPI.

#### Todos
- [X] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR

Feature #227

### CODE REVIEW

**Comment by biship** [comment]
Hay, this is exactly what I suggested in Discord last week! Nice, thanks - those of us that properly tag using Picard appreciate this. Will it be invoked both when importing (after a download) and when adding with folder import? In the case of downloading, what happens if it turns out the artist isn't what was requested - imported anyway & a new artist is created?

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 1]
```diff
@@ -234,7 +235,10 @@ public LocalTrack GetLocalTrack(string filename, Artist artist, ParsedTrackInfo
                 parsedTrackInfo = Parser.ParseMusicPath(filename);
             }
 
-            if (parsedTrackInfo == null || parsedTrackInfo.AlbumTitle.IsNullOrWhiteSpace())
+            if (parsedTrackInfo != null && parsedTrackInfo.ReleaseMBId.IsNotNullOrWhiteSpace())
+            {
+                album = _albumService.FindAlbumByRelease(parsedTrackInfo.ReleaseMBId);
```
[the remark this answers is not in the document]
Comment (majora2007) [reply src/NzbDrone.Core/Parser/ParsingService.cs]: Moved the code, but discovered another bug during testing when an AlbumTitle is empty from IDv3 tags. Will investigate a solution for that in this PR as well.

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 2]
```diff
@@ -244,9 +243,16 @@ public LocalTrack GetLocalTrack(string filename, Artist artist, ParsedTrackInfo
                 return null;
             }
 
+            /*if (parsedTrackInfo.ReleaseMBId.IsNotNullOrWhiteSpace())
+            {
+                album = _albumService.FindAlbumByRelease(parsedTrackInfo.ReleaseMBId);
+                parsedTrackInfo.AlbumTitle = album.Title;
+            }*/
```
Comment (Qstick) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Clean out commented code

File: src/NzbDrone.Core/Parser/Parser.cs [hunk 3]
```diff
@@ -641,6 +649,8 @@ private static ParsedTrackInfo ParseAudioTags(string path)
                 ArtistTitleInfo = artistTitleInfo,
                 Title = trackTitle
             };
+
+
```
Comment (Qstick) [inline src/NzbDrone.Core/Parser/Parser.cs]: Remove extra lines

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 4]
```diff
@@ -244,9 +243,16 @@ public LocalTrack GetLocalTrack(string filename, Artist artist, ParsedTrackInfo
...
+
             var tracks = GetTracks(artist, parsedTrackInfo);
-            //var album = _albumService.FindByTitle(artist.Id, parsedTrackInfo.AlbumTitle);
-            var album = tracks.FirstOrDefault()?.Album;
+            album = tracks.FirstOrDefault()?.Album;
+
+
```
Comment (Qstick) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Remove extra lines

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 5]
```diff
@@ -221,20 +221,19 @@ public LocalTrack GetLocalTrack(string filename, Artist artist)
...
-            }
-
-            else
+            } else
             {
                 parsedTrackInfo = Parser.ParseMusicPath(filename);
             }
 
-            if (parsedTrackInfo == null || parsedTrackInfo.AlbumTitle.IsNullOrWhiteSpace())
+            if ((parsedTrackInfo == null || parsedTrackInfo.AlbumTitle.IsNullOrWhiteSpace()) && parsedTrackInfo.ReleaseMBId.IsNullOrWhiteSpace())
```
Comment (Qstick) [inline src/NzbDrone.Core/Parser/ParsingService.cs]: Parenthesis are off, should be `if (parsedTrackInfo == null || (parsedTrackInfo.AlbumTitle.IsNullOrWhiteSpace() && parsedTrackInfo.ReleaseMBId.IsNullOrWhiteSpace()))`

We should always kick out if `parsedTrackInfo == null`, or we should also kick out if `AlbumTitle` and `ReleaseMBId`

File: src/NzbDrone.Core.Test/MusicTests/AlbumRepositoryTests/AlbumRepositoryFixture.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+            album.Artist = artist;
+            album.AlbumType = "";
+
+            album.Releases.Add(release);
+
+            var albumRepo = Mocker.Resolve<AlbumRepository>();
+
+            albumRepo.Insert(album);
+
+            var builtAlbum = Builder<Album>.CreateNew().BuildNew();
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/MusicTests/AlbumRepositoryTests/AlbumRepositoryFixture.cs]: `builtAlbum` variable never used

File: src/NzbDrone.Core/Music/AlbumRepository.cs [hunk 7]
```diff
@@ -301,5 +302,13 @@ public Album FindByArtistAndName(string artistName, string cleanTitle)
                         .Where<Artist>(artist => artist.CleanName == cleanArtistName)
                         .SingleOrDefault(album => album.CleanTitle == cleanTitle);
         }
+
+        public Album FindAlbumByRelease(string releaseId)
+        {
+            string query = "SELECT * FROM Albums " +
+                         "WHERE Releases LIKE '%" + releaseId + "%'";
+
+            return Query.QueryText(query).FirstOrDefault();
```
Comment (Qstick) [inline src/NzbDrone.Core/Music/AlbumRepository.cs]: This function will also match on partial release Ids. May be better off to do with all with one LINQ statement: `return Query.FirstOrDefault(e => e.Releases.Any(r => r.Id == releaseId));`
